src/mscorlib/shared/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -177,20 +189,29 @@ public override void Return(T[] array, bool clearArray = false)
                 // if there was a previous one there, push that to the global stack.  This
                 // helps to keep LIFO access such that the most recently pushed stack will
                 // be in TLS and the first to be popped next.
-                T[][] tlsBuckets = t_tlsBuckets;
+                Bucket[] tlsBuckets = t_tlsBuckets;
                 if (tlsBuckets == null)
                 {
-                    t_tlsBuckets = tlsBuckets = new T[NumBuckets][];
-                    tlsBuckets[bucketIndex] = array;
+                    t_tlsBuckets = tlsBuckets = new Bucket[NumBuckets];
+                    tlsBuckets[bucketIndex].Buffer = array;
+                    if (TrimBuffers)
+                    {
+                        s_allTlsBuckets.Add(tlsBuckets, null);
+                        Gen2GcCallback.Register(Gen2GcCallbackFunc, this);
```

Comment:
This code  path is called once for every new thread.   

However Gen2GcCallback.Register(Gen2GcCallbackFunc, this) should only be called once per Pool.    Thus you get too many callbacks when a Gen 2 happens).  

The easiest fix (assuming you want to be lazy about turning on the trimming) is to have a  flag that says it is initialized, however you have to do that with an (interlocked) operations to insure you only set it exactly once (this is a is not the common path, so you don't need to do a double-check optimization here).

Reply:
Looking...

Reply:
Added a flag and Interlocked